### DISCUSSION TITLE
Sitemap.ts - Error: Unsupported Server Component type: Null - after latest release

### DISCUSSION THREAD

**Discussion by Galanthus** [vercel/next.js#52699] [Help]
### Summary

Having a sitemap.xml which created by adding **sitemap.xl/route.ts** inside the app router is returning errors after the latest release. 

The error:

```
^[Warning: React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.
Warning: React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: null.
Error: Unsupported Server Component type: Null
    at attemptResolveElement (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1421:19)
    at resolveModelToJSON (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1627:41)
    at Object.toJSON (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1169:28)
    at stringify (<anonymous>)
    at processModelChunk (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:2116:24)
    at retryTask (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1882:38)
    at performWork (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1920:21)
    at Timeout.eval [as _onTimeout] (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1429:28)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7) {
  digest: '135229182'
}
- error node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js (1684:8) @ attemptResolveElement
- error Error: Unsupported Server Component type: Null
    at stringify (<anonymous>)
null
- error node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js (1684:8) @ attemptResolveElement
- error Error: Unsupported Server Component type: Null
    at attemptResolveElement (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1421:19)
    at resolveModelToJSON (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1627:41)
    at Object.toJSON (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1169:28)
    at stringify (<anonymous>)
    at processModelChunk (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:2116:24)
    at retryTask (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1882:38)
    at performWork (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1920:21)
    at Timeout.eval [as _onTimeout] (webpack-internal:///(sc_server)/./node_modules/next/dist/compiled/react-server-dom-webpack/cjs/react-server-dom-webpack-server.edge.development.js:1429:28)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7) {
  digest: '135229182'
}
null
```

Reverting back to the previous version doesn't show errors. Having an empty file with a simple return or console log still returns the same issue.

Must be some issue with the latest version.

```
import { SiteSettings } from "@/models/documents/site-settings"
import { SitemapRoutes } from "@/models/sitemap-routes"

import { getSitemapRoutes, getSiteSettings } from "@/lib/sanity/sanity.helpers"

export async function GET(): Promise<Response> {
  const routesPromise: Promise<SitemapRoutes[]> = getSitemapRoutes()
  const settingsPromise: Promise<SiteSettings> = getSiteSettings()

  const [routes, settings] = await Promise.all([routesPromise, settingsPromise])

  const { website_url } = settings
  const websiteUrl = website_url || "http://localhost"

  const slugs = routes.map((route) => {
    const { slug, updatedAt } = route

    const url = `${websiteUrl}${slug === "frontpage" ? "" : `/${slug}`}`

    return `
      <url>
        <loc>${url}</loc>
        <lastmod>${updatedAt}</lastmod>
      </url>
    `
  })

  const sitemapXml = `<?xml version="1.0" encoding="UTF-8" ?>
    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
      ${slugs.join("\n")}
    </urlset>`

  const headers = new Headers()
  headers.set("Content-Type", "application/xml")

  return new Response(sitemapXml, {
    headers
  })
}

export const revalidate = 60
```

**Comment by icyJoseph** [comment]
Is this in a route `app/sitemap.ts`? Next.js added support to handle sitemap.ts file at the root, and that one should either be XML or a function that returns a list with url and dates. 

- https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generate-a-sitemap

I guess it is colliding with your sitemap path. 

Could you confirm this?

**Reply by Galanthus** [reply]
Thank you. Actually I tried the same with sitemap.ts inside the app root but same result. When the sitemap.ts example is copied from its official website, I would still get the same errors. This doesn’t happen to the previous Next version. The reason why I am using the xml file instead is because revalidate doesn’t work in production only on build time. 

Can you try if you create sitemap.ts or XML with the latest Next version if you get the same console warnings?

**Reply by icyJoseph** [reply]
In what next.js version is this happening? I helped someone with a sitemap issue a couple of days ago, and had no issues.

**Reply by Galanthus** [reply]
Version **13.14.10** the one before didn't show these errors. 

This is my sitemap.ts which doesn't work or revalidate in production not sure if I am missing something, but XML supports revalidating after a time.

```
// Currently not supported with revalidate... only generates on build.

import { MetadataRoute } from "next"

import { SiteSettings } from "@/models/documents/site-settings"
import { SitemapRoutes } from "@/models/sitemap-routes"

import { getSitemapRoutes, getSiteSettings } from "@/lib/sanity/sanity.helpers"

export default async function sitemap(): Promise<MetadataRoute.Sitemap> {
  const routesPromise: Promise<SitemapRoutes[]> = getSitemapRoutes()
  const settingsPromise: Promise<SiteSettings> = getSiteSettings()

  const [routes, settings] = await Promise.all([routesPromise, settingsPromise])

  const { website_url } = settings

  const websiteUrl = website_url || "http://localhost"

  const slugs = routes.map((route) => {
    const { slug, updatedAt } = route

    const url = slug === "frontpage" ? `${websiteUrl}/` : `${websiteUrl}/${slug}`

    return {
      url,
      lastModified: updatedAt
    }
  })

  return [...slugs]
}
```

And theme /app/(website)/sitemap.xml/route.ts:

```
import { SiteSettings } from "@/models/documents/site-settings"
import { SitemapRoutes } from "@/models/sitemap-routes"

import { getSitemapRoutes, getSiteSettings } from "@/lib/sanity/sanity.helpers"

export async function GET(): Promise<Response> {
  const routesPromise: Promise<SitemapRoutes[]> = getSitemapRoutes()
  const settingsPromise: Promise<SiteSettings> = getSiteSettings()

  const [routes, settings] = await Promise.all([routesPromise, settingsPromise])

  const { website_url } = settings
  const websiteUrl = website_url || "http://localhost"

  const slugs = routes.map((route) => {
    const { slug, updatedAt } = route

    const url = `${websiteUrl}${slug === "frontpage" ? "" : `/${slug}`}`

    return `
      <url>
        <loc>${url}</loc>
        <lastmod>${updatedAt}</lastmod>
      </url>
    `
  })

  const sitemapXml = `<?xml version="1.0" encoding="UTF-8" ?>
    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
      ${slugs.join("\n")}
    </urlset>`

  const headers = new Headers()
  headers.set("Content-Type", "application/xml")

  return new Response(sitemapXml, {
    headers
  })
}

export const revalidate = 60
```

Maybe you see something I don't.

**Reply by icyJoseph** [reply]
I have started a `create-next-app` and made a `app/sitemap.xml/route.ts` path on it, and pasted this code:

```ts
export const dynamic = "force-dynamic";

export async function GET() {
  const sitemapResponse = await fetch("https://www.sitemaps.org/sitemap.xml");

  const xml = await sitemapResponse.text();

  const response = new Response(xml, {
    status: 200,
    statusText: "ok",
  });

  response.headers.append("content-type", "text/xml");

  return response;
}
```

And then this request:

```http
###
GET http://localhost:3000/sitemap.xml
```

Works without issue. In dev, and prod builds.

I'll try now with grouped routes.

**Reply by icyJoseph** [reply]
Also:

```console
yarn why next
yarn why v1.22.19
[1/4] 🤔  Why do we have the module "next"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "next@13.4.10"
```

**Reply by icyJoseph** [reply]
I think we probably are not on the same page. Could you try to setup a minimal reproduction repository? If there's indeed a bug, we'll need that anyway to be able to get it fixed quick.

https://github.com/vercel/next.js/assets/21013447/0b09537c-44cb-4393-b97a-5dd6adc026fa

**Reply by Galanthus** [reply]
Very interesting! I only see it in my vscode editor console log. How can I further debug? Because the only time I get these errors is when I try to reach the sitemap.xml in my browser. I will try that but I am pretty
Sure it’s something inside
The latest version. 

I will record my screen share in a sec

**Reply by icyJoseph** [reply]
It also works on my browser, I see the XML printout.

![Screenshot 2023-07-15 at 10 59 48](https://github.com/vercel/next.js/assets/21013447/f7f2c2f6-2a58-4286-96ba-9d4dff99ec10)

In VSCode either, [REST Client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client), or [Thunder Client](https://marketplace.visualstudio.com/items?itemName=rangav.vscode-thunder-client), are good options to see how non-browser clients/entities see about your app. 

Of course, there's CLI tools, such as cURL, that are perhaps better, or perhaps a full application such as Postman.

**Reply by Galanthus** [reply]
I will give it a try the error is to generic and doesn’t say what and where the issue is located. Thought that it’s my code for the sitemap generation but still getting the error.

Will output the result here. Maybe it conflicts
With other node packages? Not sure if that is possible.

**After debugging and removing all app files this is the one causing it:**

**app/(website)/[slug]/page.tsx**

```
import { draftMode } from "next/headers"
import { notFound, redirect } from "next/navigation"

import { PreviewProvider } from "@/context"

import { Page } from "@/models/documents/page"
import { PageRoute } from "@/models/page-route"

import { allPagesSlug, pageQuery } from "@/lib/queries"
import { cachedClient } from "@/lib/sanity/sanity.client"
import { getCachedClient } from "@/lib/sanity/sanity.getClient"
import { getPageMeta } from "@/lib/sanity/sanity.helpers"

import { PageLayout, PagePreviewLayout } from "@/components/layout"

export const generateMetadata = async ({ params }: { params: { slug: string } }) => {
  const { slug } = params
  const page = await getPageMeta({ type: "page", slug })

  return page
}

export const generateStaticParams = async () => {
  const pages = await cachedClient<Page[]>(allPagesSlug)

  return pages
}

export const revalidate = 60

const IndexPage = async ({ params }: PageRoute) => {
  const preview = draftMode().isEnabled ? { token: process.env.SANITY_API_READ_TOKEN } : undefined

  const page = await getCachedClient(preview)(pageQuery, params)

  if (params.slug === "rechtsgebied") {
    redirect("/rechtsgebieden")
  }

  if (!page) {
    notFound()
  }

  if (preview && preview.token) {
    return (
      <PreviewProvider token={preview.token}>
        <PagePreviewLayout page={page} />
      </PreviewProvider>
    )
  }

  return <PageLayout page={page} />
}

export default IndexPage

```
Could you please, be so kind if you see something weird? it's 100% this file commenting the complete code doesn't show any errors. 

**/app/(website)/[slug]/page.tsx** and the **/app/(website)/sitemap.xml/route.ts** conflict.

**Reply by Galanthus** [reply]
When I open the http://localhost:3000/sitemap.xml it returns the errors inside my terminal console tab on vscode. But running the URL with Thunder client doesn't show any errors at all.

Screenshot:
https://i.ibb.co/y5RVxJM/Screenshot-2023-07-15-at-11-23-03.png

After removing all the app folders and running it doesn't show any errors.. the question is how can I debug where this error is located? Are there any tools, plugins, extensions maybe? Because removing all files from a pretty large application is not the solution for debugging.

**Reply by icyJoseph** [reply]
mmm I wonder if there a layout wrapper might be bugging out? do you have a layout component in `app/(website)`?

**Reply by icyJoseph** [reply]
I see, well I can't break it on my side... I made a page with slug param, revalidate, generateStaticParams, and notFound, yet no error.

What's up with PageLayout or any of the other components? 

if this page was just:

```tsx
const IndexPage = async ({ params }: PageRoute) => {
  return <div>Hello world</div>
}
```

Does it still trigger the issue?

**Reply by Galanthus** [reply]
Thanks friend.

```
 if (!page) {
    notFound()
  }
```

commenting this code removes the error.

So, the code causing this is coming from this code:

```
import { cache } from "react"

import type { SanityClient } from "@sanity/client"
import { createClient } from "next-sanity"

import { apiVersion, dataset, projectId, useCdn } from "./sanity.env"

export const getClient = (preview?: { token?: string }): SanityClient => {
  const client = createClient({
    projectId,
    dataset,
    apiVersion,
    useCdn,
    perspective: "published"
  })

  if (preview) {
    if (!preview.token) {
      throw new Error("You must provide a token to preview drafts")
    }
    return client.withConfig({
      token: preview.token,
      useCdn: false,
      ignoreBrowserTokenWarning: true,
      perspective: "previewDrafts"
    })
  }
  return client
}

const getCachedClient = (preview?: { token?: string }) => {
  const client = getClient(preview)

  return cache(client.fetch.bind(client))
}

export { getCachedClient }
```

**Reply by Galanthus** [reply]
Stripped the complete code and the error shows only if the notFound is checking for a page.

it's the **`NotFound()`** when I replace it with **`redirect("/")`** the error is gone.

**Reply by icyJoseph** [reply]
Aha, so is it trying to generate that page for sitemap? And that trashes it? Still, why would it work for non browsers. So many questions 😂 but now we know the lines that trigger the error.

**Reply by Galanthus** [reply]
I have multiple **`notFound()`** on other routes where I have the page.tsx but this only happend in the root **[slug]/page.tsx** So can it be related to the latest next? Could you also try if you create the same folder like I did and add a function NotFound to it?

**Reply by icyJoseph** [reply]
I had that. It was a page with nothing but a notFound call.

I'd say, that now it is better if you setup a repository with the error/misbehavior, so that we can draw some conclusions, and possibly open an issue with a reproduction repository.

**Comment by Galanthus** [comment]
> It also works on my browser, I see the XML printout.
> 
> ![Screenshot 2023-07-15 at 10 59 48](https://user-images.githubusercontent.com/21013447/253730928-f7f2c2f6-2a58-4286-96ba-9d4dff99ec10.png)
> 
> In VSCode either, [REST Client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client), or [Thunder Client](https://marketplace.visualstudio.com/items?itemName=rangav.vscode-thunder-client), are good options to see how non-browser clients/entities see about your app.
> 
> Of course, there's CLI tools, such as cURL, that are perhaps better, or perhaps a full application such as Postman.



> I will give it a try the error is to generic and doesn’t say what and where the issue is located. Thought that it’s my code for the sitemap generation but still getting the error.
> 
> Will output the result here. Maybe it conflicts With other node packages? Not sure if that is possible.
> 
> **After debugging and removing all app files this is the one causing it:**
> 
> **app/(website)/[slug]/page.tsx**
> 
> #import { draftMode } from "next/headers" import { notFound, redirect } from "next/navigation"
> 
> import { PreviewProvider } from "@/context"
> 
> import { Page } from "@/models/documents/page" import { PageRoute } from "@/models/page-route"
> 
> import { allPagesSlug, pageQuery } from "@/lib/queries" import { cachedClient } from "@/lib/sanity/sanity.client" import { getCachedClient } from "@/lib/sanity/sanity.getClient" import { getPageMeta } from "@/lib/sanity/sanity.helpers"
> 
> import { PageLayout, PagePreviewLayout } from "@/components/layout"
> 
> export const generateMetadata = async ({ params }: { params: { slug: string } }) => { const { slug } = params const page = await getPageMeta({ type: "page", slug })
> 
> return page }
> 
> export const generateStaticParams = async () => { const pages = await cachedClient<Page[]>(allPagesSlug)
> 
> return pages }
> 
> export const revalidate = 60
> 
> const IndexPage = async ({ params }: PageRoute) => { const preview = draftMode().isEnabled ? { token: process.env.SANITY_API_READ_TOKEN } : undefined
> 
> const page = await getCachedClient(preview)(pageQuery, params)
> 
> if (params.slug === "rechtsgebied") { redirect("/rechtsgebieden") }
> 
> if (!page) { notFound() }
> 
> if (preview && preview.token) { return ( ) }
> 
> return }
> 
> export default IndexPage
> 
> Could you please, be so kind if you see something weird?



> if (!page) {
>     notFound()
>   }



> I see, well I can't break it on my side... I made a page with slug param, revalidate, generateStaticParams, and notFound, yet no error.
> 
> What's up with PageLayout or any of the other components?
> 
> if this page was just:
> 
> ```tsx
> const IndexPage = async ({ params }: PageRoute) => {
>   return <div>Hello world</div>
> }
> ```
> 
> Does it still trigger the issue?



> Aha, so is it trying to generate that page for sitemap? And that trashes it? Still, why would it work for non browsers. So many questions 😂 but now we know the lines that trigger the error.

It has nothing to do with the sitemap it works just fine. It's the headless CMS Sanity which is the main cause. Reverting back to previous Next version works just fine. It's complaining about if the page is NULL normally, it will return the not found page. In previous and latest version when I console log a page that doesn't exist, one returns the error page and console log null, the latest version doesn't redirect but does return the same console error null.

Btw, Does the latest sitemap.ts support revalidate? I like it it's cleaner code, but it seems like you can't revalidate in production. Thanks in advance!

**Comment by markudevelop** [comment]
I also get this and it has nothing to do with sitemap.xml any route with slugs using notFound will return this error 

e.g. app/[locale]/blog/[slug]/page.tsx -> notFound()

**Reply by icyJoseph** [reply]
Got a repository we could explore and possibly use to open up a bug report? If there isn't one already

**Reply by markudevelop** [reply]
@icyJoseph 
Here you go:

https://stackblitz.com/edit/stackblitz-starters-mtdjqq?file=package.json
(visit the following path /en/about/test)

I've tested using the latest next.js version if I change to an older one ("next": "^13.4.6-canary.3") works fine!

**Reply by Galanthus** [reply]
Thank you for sharing.

**Reply by icyJoseph** [reply]
I think in this case the problem is the missing root layout, which is required:

- https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#root-layout-required

**Reply by stijnbe** [reply]
Moving the RootLayout back to the /app directory fixed the problem. However in the internationalization guide in the docs there is a suggestion to move the root layout to app/[lang]/layout.js
https://nextjs.org/docs/app/building-your-application/routing/internationalization

**Reply by YuchanJeong** [reply]
I encountered a similar issue with the following file structure:

```
app/
  - [lng]/
    - page.tsx
    - layout.tsx
```

After changing it to the following structure, the problem was resolved. While @stijnbe's suggestion is worth considering, for now, I will proceed with the new structure:

```
app/
  - [lng]
  - layout.ts
```
